feat(aws-ecs-builder): add public ip support

misterjoshua opened the pull request:
This change introduces `AssignPublicIpExtension`. This feature assigns a public IP address to Fargate tasks. The user may also opt to have the tasks automatically register their public IPs in a Route53 Hosted Zone.

Closes #10644

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 45d8975bb1868b699777b8244c1cdc01f2c9a093
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3855f626-8703-45c5-a316-6f2fc509fd12%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: df2b159e10ee7f7b89e811651597829d69a3f415
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1ba89be3-7a7b-47a2-b554-65b494e13f78%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nathanpeck commented:
Thanks so much for the contribution! Two things I'd like to cover before merge:

**First,** from the docs: "The awsvpc network mode does not provide task ENIs with public IP addresses for tasks that use the EC2 launch type." (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-networking.html)

So this extension needs to do some checking in the prehook to make sure the parent environment is actually using Fargate capacity. Could you add a test for an EC2 environment and verify what happens? We may need to throw an exception if the user tries to assign a public IP in an EC2 instance, especially if that EC2 is in a private subnet anyway.

Here is another test that creates an EC2 environment for reference: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk-containers/ecs-service-extensions/test/test.environment.ts#L149-L160

**Second**, I'd like to discuss what this extension is accomplishing to see if its mission can be better reflected in terms of a usecase instead of just the name of the service object setting that it modifies. What is the underlying use case you are targeting with the public IP address feature? Imagine a person who chooses to add this extension. What is the problem they were trying to solve that motivated them to add the extension to their service? Because maybe that problem can be better reflected in the name of the extension and its features.

For example it sounded from issue #10644 that maybe the goal was to allow users to do direct inbound requests to the task using its IP with no load balancer needed. If so then perhaps we eventually would want to have something like this: `publicDnsHostnameExtension` which not only enables public IP addresses on the task, but also creates a public facing Route53 record pointing at the IP address, and opens the traffic port of the service's security group to inbound traffic from anywhere, and a `CfnOutput` for getting the public hostname of the task to point traffic at. I know that this broadens the scope of the extension a lot and makes it harder to implement but it would totally be worth it in the long run.

I think this extension is super interesting, and thank you again for your contribution! I'd just like to make sure it considers the EC2 environment as well (even if that means it just throws an exception that says this only works with Fargate). And perhaps the name and features of the extension could target a specific use case for the end user, rather than just mutating a single setting.

nathanpeck commented:
Adding some further notes: according to this issue https://github.com/aws/containers-roadmap/issues/737 and some of my other anecdotal conversations with folks a lot of people who want direct public IP addressing for their tasks use a container dependency ordering sidecar which runs before their application container, and self registers the task's public IP into Route53. That sidecar runs to completion and exits, and then the application is setup with the container dependency condition set to `SUCCESS` for the sidecar so that it runs after the sidecar runs.

So in that case a full flow for this extension could look like:

- Update service props to have a public IP
- Open the security group for the service's traffic port to traffic from anywhere (or optionally it could be still locked down and you have to use `connectTo` to open access only from other connectable sources)
- Add a sidecar to the task definition which just runs a small bash script that queries the [ECS task metadata endpoint](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html) to get the tasks own IP address and registers that IP address into a Route53 record
- Set a container dependency between the app container and the sidecar so that the app container runs after the sidecar
- Add a CloudFormation output with the DNS name from Route53 so that there is a way to get the domain name of the task.

There are still a few things to consider though:

- What about services that launch more than 1 task?
- What about cases where the application crashes and the Route 53 record is left hanging? (Run another sidecar after main application container completion, to clean things up perhaps?)

misterjoshua commented:
> Thanks so much for the contribution! Two things I'd like to cover before merge:
> 
> **First,** from the docs: "The awsvpc network mode does not provide task ENIs with public IP addresses for tasks that use the EC2 launch type." (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-networking.html)
> 
> So this extension needs to do some checking in the prehook to make sure the parent environment is actually using Fargate capacity. Could you add a test for an EC2 environment and verify what happens? We may need to throw an exception if the user tries to assign a public IP in an EC2 instance, especially if that EC2 is in a private subnet anyway.
> 
> Here is another test that creates an EC2 environment for reference: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk-containers/ecs-service-extensions/test/test.environment.ts#L149-L160

No problemo. I've made this change and pushed it up. I'll address your second talking point shortly. :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2817e9c90c8ec2ab8a42bf19482f4fb5fde5b801
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fcfff541-a0ab-4af0-82e5-45f88372d0c4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
> **Second**, I'd like to discuss what this extension is accomplishing to see if its mission can be better reflected in terms of a usecase instead of just the name of the service object setting that it modifies. What is the underlying use case you are targeting with the public IP address feature? Imagine a person who chooses to add this extension. What is the problem they were trying to solve that motivated them to add the extension to their service? Because maybe that problem can be better reflected in the name of the extension and its features.
> 
> For example it sounded from issue #10644 that maybe the goal was to allow users to do direct inbound requests to the task using its IP with no load balancer needed. If so then perhaps we eventually would want to have something like this: `publicDnsHostnameExtension` which not only enables public IP addresses on the task, but also creates a public facing Route53 record pointing at the IP address, and opens the traffic port of the service's security group to inbound traffic from anywhere, and a `CfnOutput` for getting the public hostname of the task to point traffic at. I know that this broadens the scope of the extension a lot and makes it harder to implement but it would totally be worth it in the long run.
> 
> I think this extension is super interesting, and thank you again for your contribution! I'd just like to make sure it considers the EC2 environment as well (even if that means it just throws an exception that says this only works with Fargate). And perhaps the name and features of the extension could target a specific use case for the end user, rather than just mutating a single setting.

@nathanpeck Yes, you're right that I don't want a load balancer in this case. Like in aws/containers-roadmap#737, I have cost concerns about spinning up load balancers for these services. So, it makes a lot more sense from where I'm sitting to expose these tasks through public IP addresses so that the user can directly interact with them. DNS support would absolutely help me.

> Adding some further notes: according to this issue [aws/containers-roadmap#737](https://github.com/aws/containers-roadmap/issues/737) and some of my other anecdotal conversations with folks a lot of people who want direct public IP addressing for their tasks use a container dependency ordering sidecar which runs before their application container, and self registers the task's public IP into Route53. That sidecar runs to completion and exits, and then the application is setup with the container dependency condition set to `SUCCESS` for the sidecar so that it runs after the sidecar runs.
> 
> So in that case a full flow for this extension could look like:
> 
> * Update service props to have a public IP
> * Open the security group for the service's traffic port to traffic from anywhere (or optionally it could be still locked down and you have to use `connectTo` to open access only from other connectable sources)
> * Add a sidecar to the task definition which just runs a small bash script that queries the [ECS task metadata endpoint](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html) to get the tasks own IP address and registers that IP address into a Route53 record
> * Set a container dependency between the app container and the sidecar so that the app container runs after the sidecar
> * Add a CloudFormation output with the DNS name from Route53 so that there is a way to get the domain name of the task.
> 
> There are still a few things to consider though:
> 
> * What about services that launch more than 1 task?
> * What about cases where the application crashes and the Route 53 record is left hanging? (Run another sidecar after main application container completion, to clean things up perhaps?)

I like this sidecar approach, but I wonder if we could better solve the dangling record problem by implementing the DNS component using Lambda and EventBridge. For instance:

* Create an EventBridge rule to trigger a Lambda Function on ECS Task state changes.
* When candidate tasks transition to `RUNNING`, create/update Route 53 records.
* When candidate tasks transition to `STOPPED`, remove/update the Route 53 records.

When there is more than one task per record, perhaps we could have the lambda handle merging. I suspect that we'll need to store the task's IP so that we can correctly remove the IP from the merged recordset. Perhaps we can store the IPs as a tag on the task when the task transitions to `RUNNING`. Then we can look for that tag when handling a `STOPPED` transition.

nathanpeck commented:
> I wonder if we could better solve the dangling record problem by implementing the DNS component using Lambda and EventBridge. 

So I don't remember if the EventBridge notification actually has the task IP in it or not. Would need to double check on that. Worst case scenario the Lambda can always query the task details if needed. I do think EventBridge would be a more durable and reliable mechanism for fetching the task IP and putting it in the DNS record. One more thing to consider is what happens if multiple tasks launch at the same time and multiple Lambda invokes are doing parallel `ChangeResourceRecordSets` on the same multivalue A record. All stuff that can be solved though

If you are up to implementing it using EventBridge that sounds good to me! This has a lot of potential to turn into one of the first extensions which doesn't only provide a task definition focused feature, but actually deploys supporting tooling around the task as well.

misterjoshua commented:
> > I wonder if we could better solve the dangling record problem by implementing the DNS component using Lambda and EventBridge.
> 
> So I don't remember if the EventBridge notification actually has the task IP in it or not. Would need to double check on that. Worst case scenario the Lambda can always query the task details if needed. I do think EventBridge would be a more durable and reliable mechanism for fetching the task IP and putting it in the DNS record. One more thing to consider is what happens if multiple tasks launch at the same time and multiple Lambda invokes are doing parallel `ChangeResourceRecordSets` on the same multivalue A record. All stuff that can be solved though
> 
> If you are up to implementing it using EventBridge that sounds good to me! This has a lot of potential to turn into one of the first extensions which doesn't only provide a task definition focused feature, but actually deploys supporting tooling around the task as well.

I'll take a crack at this and push up my minimum viable. There's definitely a distributed data problem in play here, but as you say it's not an insurmountable problem. :)

misterjoshua commented:
I put in a little time tonight. I have EventBridge feeding `RUNNING` and `STOPPED` tasks transitions into a queue and a lambda trigger with reserved concurrent executions of 1. The lambda is doing a great job of fetching the attached ENI public IPs and adding ips to the record set.

However, my idea of tagging ECS tasks with IP information isn't going to work well. This approach will only work when long-format arns are enabled on the account. I may need to store the ips elsewhere, like in a DynamoDB table. I'll dig a little deeper later.

nathanpeck commented:
Awesome! Yeah I was thinking a DynamoDB table would be best anyway because
you could use the optimistic locking with version number approach (
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.OptimisticLocking.html)
to protect against race conditions of multiple lambda invokes updating
Route 53 at the same time.

If it is provisioned in the "pay per request" mode it would be super cheap
anyway (only used and only charging when a task is stopped or launched).

On Sun, Oct 4, 2020 at 3:02 AM Josh Kellendonk <notifications@github.com>
wrote:

> I put in a little time tonight. I have EventBridge feeding RUNNING and
> STOPPED tasks transitions into a queue and a lambda trigger with reserved
> concurrent executions of 1. The lambda is doing a great job of fetching the
> attached ENI public IPs and adding ips to the record set.
>
> However, my idea of tagging ECS tasks with IP information isn't going to
> work well. This approach will only work when long-format arns are enabled
> on the account. I may need to store the ips elsewhere, like in a DynamoDB
> table. I'll dig a little deeper later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aws/aws-cdk/pull/10646#issuecomment-703212976>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2VN6QFT6QZHY242ZA3DCLSJAMZHANCNFSM4SBGBF6A>
> .
>

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0006e2ca7cfeae7a0b9726c64795d135374e7acd
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e38ac1ab-164f-4623-87be-d333b6e251c1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@nathanpeck I put in a little more effort tonight. I've rebased away much of my junk from yesterday and pushed up my WIP, in case you're interested in taking a peek.

The architecture is like this so far: 

```
RUNNING or STOPPED transition event
 ↓
SQS
 ↓
Lambda trigger (unlimited concurrency)
 ↓
Optimistically locked put to DynamoDB.
```

The DynamoDB record holds the list of IPs as well as task information. I'm also keeping running task info so I know what IP to remove once the ENIs are gone. I also keep stopped task tombstones in case I get some out-of-order STOPPED then RUNNING events. I can't imagine will ever happen, but you can never know with a standard queue. The lambda expunges the tombstones after a suitable amount of time has passed. I've included a trimmed up example of the DynamoDB record below.

I had hoped that I could drive the lambda with larger batch sizes to reduce the hits on route53. It would have reduced the number of `ecs:DescribeNetworkInterfaces` calls, but it looks like I'm eating batch-size-1 receives. Open to suggestions.

I'll pick up from where I left off as soon as I can find a chance.

**Trimmed up example DynamoDB record**

```json
{
  "hosted_zone_id_record_name": "SOMETHING#test-record.myexample.com",
  "ipv4s": [
    "1.1.1.80",
    "1.1.1.198",
    "1.1.1.30",
    "1.1.1.149",
    "1.1.1.132"
  ],
  "task_info": {
    "arn:aws:ecs:::task/203820d1-5bb8-4629-8c77-5b3f9a45e2cc": {
      "enis": [
        {
          "eni_id": "eni-0ec6bfc7c387dbc03",
          "public_ipv4": "1.1.1.147"
        }
      ],
      "stopped_datetime": "2020-10-05T04:10:03.260863",
      "task_arn": "arn:aws:ecs:::task/203820d1-5bb8-4629-8c77-5b3f9a45e2cc"
    },
    "arn:aws:ecs:::task/23eb3e7a-a8ea-4772-8fe4-e182f6a6b132": {
      "enis": [
        {
          "eni_id": "eni-09521754827aad43d",
          "public_ipv4": "1.1.1.231"
        }
      ],
      "stopped_datetime": "2020-10-05T04:10:19.039618",
      "task_arn": "arn:aws:ecs:::task/23eb3e7a-a8ea-4772-8fe4-e182f6a6b132"
    },
    "arn:aws:ecs:::task/4d21810a-40ef-4a2f-bb10-6241fb426d6f": {
      "enis": [
        {
          "eni_id": "eni-0b265464f8c72a9ba",
          "public_ipv4": "1.1.1.238"
        }
      ],
      "stopped_datetime": "2020-10-05T04:09:54.865707",
      "task_arn": "arn:aws:ecs:::task/4d21810a-40ef-4a2f-bb10-6241fb426d6f"
    },
    "arn:aws:ecs:::task/50268c28-351e-4bce-bf44-7b0f35a4e3ea": {
      "enis": [
        {
          "eni_id": "eni-0d414fe62f2f634d9",
          "public_ipv4": "1.1.1.198"
        }
      ],
      "task_arn": "arn:aws:ecs:::task/50268c28-351e-4bce-bf44-7b0f35a4e3ea"
    },
    "arn:aws:ecs:::task/97729fe5-1b9c-4b53-971e-d8a6991ba4fd": {
      "enis": [
        {
          "eni_id": "eni-00ddd86d68a5b4219",
          "public_ipv4": "1.1.1.165"
        }
      ],
      "stopped_datetime": "2020-10-05T04:09:57.623562",
      "task_arn": "arn:aws:ecs:::task/97729fe5-1b9c-4b53-971e-d8a6991ba4fd"
    },
    "arn:aws:ecs:::task/af64a125-83bb-4d06-b509-c91e68d14567": {
      "enis": [
        {
          "eni_id": "eni-02635da194633f778",
          "public_ipv4": "1.1.1.149"
        }
      ],
      "task_arn": "arn:aws:ecs:::task/af64a125-83bb-4d06-b509-c91e68d14567"
    },
    "arn:aws:ecs:::task/c1f8ea42-4d05-4481-b512-b765e252e469": {
      "enis": [
        {
          "eni_id": "eni-082c03bea8dd4a210",
          "public_ipv4": "1.1.1.30"
        }
      ],
      "task_arn": "arn:aws:ecs:::task/c1f8ea42-4d05-4481-b512-b765e252e469"
    },
    "arn:aws:ecs:::task/cf649102-f27a-41ba-859d-929cf1bd1c11": {
      "enis": [
        {
          "eni_id": "eni-0c7f473a3aa1fc838",
          "public_ipv4": "1.1.1.137"
        }
      ],
      "stopped_datetime": "2020-10-05T04:09:53.265616",
      "task_arn": "arn:aws:ecs:::task/cf649102-f27a-41ba-859d-929cf1bd1c11"
    },
    "arn:aws:ecs:::task/dc60122e-855d-46f1-8a43-5de4b41cd45c": {
      "enis": [
        {
          "eni_id": "eni-0f7ca43b7f3b0b54e",
          "public_ipv4": "1.1.1.80"
        }
      ],
      "task_arn": "arn:aws:ecs:::task/dc60122e-855d-46f1-8a43-5de4b41cd45c"
    },
    "arn:aws:ecs:::task/defe75f5-0cbb-463d-a32a-764fdf06ddd3": {
      "enis": [
        {
          "eni_id": "eni-013e2cffea015a161",
          "public_ipv4": "1.1.1.132"
        }
      ],
      "task_arn": "arn:aws:ecs:::task/defe75f5-0cbb-463d-a32a-764fdf06ddd3"
    }
  },
  "version": 21
}
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 73f89f71b4db02d59ea65fe8f730f8ddf2325d4b
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=675b9fbd-9977-49cb-8677-a8b21568f344%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0173e3031f4028bd2e6e4b4c10a7c903eeb582c3
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cd9a28d5-9188-4510-9b1a-1945a0afb94c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9408ff6d5433afd3172d32e75931c250876a2ea9
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0daa7eb2-6398-4202-ae54-e247f55589c3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nathanpeck commented:
Awesome job! I'm reading through the code right now and it looks good. I want to try out the integration test on my own account and stress test it a little to see if I find any issues, but otherwise it looks great, and once I test it out we can approve for merge.

> I had hoped that I could drive the lambda with larger batch sizes to reduce the hits on route53 and the requests for ENI descriptions. But, it looks like I'm eating batch-size-1 receives. Given this, I'm not sure the SQS part is worth it unless there's an easy way to increase batch sizes that I'm not thinking of. Open to suggestions.

SQS does not have any concept of pausing to wait for a batch to fill up. It is optimized for fast delivery to an end consumer over waiting for a batch to fill up, so it will only deliver larger batches to consumers if there are actually many concurrent items being posted to the queue such that when the consumer polls for a batch there is a full batch waiting to be delivered. So you are right the SQS queue isn't adding much utility right now as it is. The queue might make sense if you were combining it with `reservedConcurrentExecutions = 1`  for the Lambda so that only one Lambda invoke could happen at a time. Might be worth an experiment to see how that works.

One alternative way that you can solve the "reduce the hits on route53 and the requests for ENI descriptions" problem is to split it into two pieces: collector and applier. The collector just collects the events into the DynamoDB record, as fast as they come in. The applier executes on a schedule or timeout (maybe once every 30 seconds or every 1 min) and does the actual update to Route 53. Of course this would cause it to be less reactive to task changes.

In my opinion this extension should probably only be used for small deployments anyway, so it hopefully won't be a problem to hit quotas, so no need to split it apart into collector/applier. I'll experiment with the integration tests to see if I can hit quotas or anything though, and based on that we can document the limits.

nathanpeck commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/lambda/lib/route53.py:
```diff
@@ -0,0 +1,91 @@
+
+    def upsert(self, locator: Route53RecordSetLocator, record_set):
+        logging.info(f'Upserting record set {record_set}')
+        self.route53_client.change_resource_record_sets(
```
I believe we need an error catch and retry here. According to the docs:

> ChangeResourceRecordSets requests
If Route 53 can't process a request before the next request arrives, it will reject subsequent requests for the same hosted zone and return an HTTP 400 error (Bad request). The response header also includes a Code element with a value of PriorRequestNotComplete and a Message element with a value of The request was rejected because Route 53 was still processing a prior request.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-requests

misterjoshua replied:
@nathanpeck Definitely worth considering. I have to perform an overhaul of this class anyway since it's not yet hooked up to the DynamoDB table info.

Otherwise, as it relates to Route53, I've been worrying a little about the potential for another race condition. It seems to me that the part that happens after the DynamoDB item is merged/received is still vulnerable to a data race. Diagram below. Given this, I wonder if I should stick with reserved concurrent executions of 1 for the lambda in question.

**Race Diagram**
Event A comes in followed by Event B. The optimistic locking for the DynamoDB portion results in both A's DDB Record and B's DDB Record being correct, even though Event A and event B happen concurrently. A's Record Set is out of date because event B happened, but it doesn't know it. Then, A and B are racing to put their version up on Route53. If A receives a delay, B may update the Record Set first, then A may overwrite B's Record Set without realizing that it's out of date.

```
Time ->

Event A -------> DDB Merge -------> A's DDB Record --------------> Backoff delay --------------> A's Record Set
     Event B -------> DDB Merge -------> B's DDB Record --------------> B's Record Set
```

nathanpeck replied:
Yep you are totally right about that race. I think a concurrency of 1 for the lambda is necessary for sure and would solve the problem. Events should just queue up in SQS and apply one at a time. In fact if multiple events had queued up while the first invocation was happening then the next batch would actually come through as multiple records instead of one record so that would help solve the problem of hitting rate limits

misterjoshua replied:
I've added the catch you mentioned in a recent commit, have added the retries by creating a helper, and dealt with the race condition we discussed by dropping to concurrency of 1.

nathanpeck commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/assign-public-ip.ts:
```diff
@@ -0,0 +1,69 @@
+
+export interface AssignPublicIpExtensionOptions {
+  /**
+   * DNS Zone to expose expose a record to. If specified, `dnsRecordName` is
```
To make the optional/required behavior a bit clearer let's put both these properties into their own interface that is optional, but this sub interface has both properties required. That way the TypeScript compiler can catch misuse rather than needing a runtime check.

```ts
new AssignPublicIpExtension({
  dns: {
    dnsZone: zone,
    dnsRecordName: name
  }
})
```

`dns` is optional, but inside `dns` both subprops are required.

misterjoshua replied:
Great idea! I'll make this fix asap

nathanpeck commented on packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/lambda/lib/route53.py:
```diff
@@ -0,0 +1,91 @@
+
+    def delete(self, locator: Route53RecordSetLocator, record_set):
+        logging.info(f'Deleting record set: {record_set}')
+        self.route53_client.change_resource_record_sets(
```
Same here, maybe a helper method with exponential backoff on the retry so that way it can handle the 5/sec default quota on Route53 API calls as well:

> Five requests per second per AWS account. If you submit more than five requests per second, Amazon Route 53 returns an HTTP 400 error (Bad request). The response header also includes a Code element with a value of Throttling and a Message element with a value of Rate exceeded.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-requests

misterjoshua replied:
I've added a catch for this in a recent commit.

misterjoshua commented:
I got another chance to work on this tonight.
- Changes from review (typescript interface, concurrency of 1, exponential backoff, handling an exception)
- A few bug fixes relating to matching ECS events (wasn't handling `lastStatus=RUNNING` `desiredStatus=STOPPED` for instance)
- Cleanup (moved some things around, added more testing, renamed)

Two glaring things remain for me:
- There are some problems deleting the stack because of dependency ordering and deleting the zone with the record set.
- I still want to get test coverage on the main part of the lambda handler, so I'm refactoring and cleaning there.

On the stack deleting problem: It's a little tough because I'm facing a race in determining the service name. I'm racing to see the first RUNNING when adding the service name to the event bridge rule, or racing to see the last (few?) STOPPED if I filter in the lambda. If I race to see the last few STOPPED, the record set dangles. I'm considering creating a custom resource to help clean out the record set. I figure that the custom resource doesn't need to know the service name, so I can work around this issue.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 79a8cf20c4e4fd4a54f326ce55248ae675c4bae4
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=61901a45-e7f9-4c6d-913a-3adcf9b840d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 62cf9e5dfb86620213888cf12ba98b293403b44a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e87338db-b97d-4db5-af0d-f3e410f2079d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c51af4e01348d88e7acea2a34eaf3e71fb79fb95
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1abfd325-a507-40b0-87e8-c7d3c3688fc5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
I took a crack at the custom resource type. At least from my testing, this seems to fix the stack deletion issue.

Given the heft of the tag-along resources for custom resource providers, I'm inclined to want to turn the custom resource bit into its own construct and add only one provider per stack.

I also need to make the manual verification of the integration test after deployment easier. I've been having fun manually turning desired tasks up to and down from hundreds and watching the record set follow suit. :p

I still feel the need to improve test coverage, but all the manual testing seems to be going great right now. I'm nearly satisfied haha.

I'll pick this up again later.

nathanpeck commented:
> Given the heft of the tag-along resources for custom resource providers, I'm inclined to want to turn the custom resource bit into its own construct and add only one provider per stack.

Yeah I ran into this for another thing too, and I was thinking as a future improvement to add the concept of "environment" scoped resources. So an extension can add a shared resource to the environment which can then be reused across multiple instantiations of the extension. I'm fine with overhead for now, and optimizing later on once we spec out how to have shared environment wide resources

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: acfa3515e0a9249222141988f3890d121a3d6895
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9fb99e54-6a90-449c-8e4b-35255046fd25%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2f74ed16e7e07557cd6abd33ba038564e8e7bbfc
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6b7a60d7-23e7-4197-8a56-c4c60b0e74fb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@nathanpeck After the most recent commits, I'm generally happy with this feature. But, admittedly, I'm a little close to the canvas. :) Anyway, the code's ready for review and testing when you get a chance. I'll keep an eye out for feedback.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9a3e109eea688bdb185cd708f374104201a2d0a7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=75caeee2-2c3d-4678-b3b7-6c10bae213f4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 113eee1a344cb4cfb0ff08a43f9d0b468e90cc0d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f790fae0-0373-424e-bb9a-1220eacbe95e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nathanpeck commented:
Awesome thanks Joshua! I'll be taking a look and adding some others from the team to review as well!

nathanpeck commented:
Okay I tried it out with some of my own domains! It works great! I only found a few issues:

- The extension does not open the security group to the public for inbound access. So the public DNS record works but end users can't reach the backends as their security group is blocking inbound traffic. This should be a simple addition in `useService`:

  ```ts
  public useService(service: ecs.Ec2Service | ecs.FargateService) {
      if (this.hasDns()) {
        new TaskRecordManager(service, 'TaskRecordManager', {
          service: service,
          dnsZone: this.dns!.zone,
          dnsRecordName: this.dns!.recordName,
        });
  
        const container = this.parentService.serviceDescription.get('service-container') as Container;
        service.connections.allowFromAnyIpv4(
          ec2.Port.tcp(container.trafficPort),
          'Accept inbound traffic on traffic port from anywhere',
        );
      }
    }
  ```

  In keeping with the spirit of default config that works out of the box, I think the extension should open the SG itself if you have specified a DNS zone
- If you change the DNS zone property after deploying the service (as in create the service with record in one DNS zone, and then change the dns zone it seems to get confused (to be expected though, tbh). I don't think this is worth investing a ton of effort into as its an extreme edge case, but we should perhaps add a comment in the doc string for the extension indicating that you will want to remove the extension completely and then readd it if you want to change the DNS zone.
- This is the most important issue. If the hosted zone has multiple records it sometimes overwrites the wrong one. See in this example the Lambda was configured to write to `name.nathanpeck.gg`

  ![Screen Shot 2020-10-09 at 3 04 55 PM](https://user-images.githubusercontent.com/3495674/95622156-e91b5780-0a40-11eb-9086-c7b06fe95915.png)

   However when it actually executed it selected another prexisting record in the same hosted zone: `stats.nathanpeck.gg` and overwrite it instead of writing to `name.nathanpeck.gg`
![Screen Shot 2020-10-09 at 3 04 42 PM](https://user-images.githubusercontent.com/3495674/95622203-00f2db80-0a41-11eb-80a1-833a6471373a.png)

  I believe that these lines right here need to be modified to select the correct matching record, rather than just the first record: https://github.com/aws/aws-cdk/pull/10646/files#diff-597cb6e28116f127342830001126eb7dR46-R48

nathanpeck commented:
![Screen Shot 2020-10-09 at 3 43 25 PM](https://user-images.githubusercontent.com/3495674/95625238-2fbf8080-0a46-11eb-91ee-1401ff8b5462.png)
Perhaps also worth pointing out this error. This is because it tried to update the wrong record though, a CNAME record instead of an A record, so it should be fixed by a fix to update the right A record.

misterjoshua commented:
> Okay I tried it out with some of my own domains! It works great! I only found a few issues:
> 
> * The extension does not open the security group to the public for inbound access. So the public DNS record works but end users can't reach the backends as their security group is blocking inbound traffic. This should be a simple addition in `useService`:
>   ```ts
>   public useService(service: ecs.Ec2Service | ecs.FargateService) {
>       if (this.hasDns()) {
>         new TaskRecordManager(service, 'TaskRecordManager', {
>           service: service,
>           dnsZone: this.dns!.zone,
>           dnsRecordName: this.dns!.recordName,
>         });
>   
>         const container = this.parentService.serviceDescription.get('service-container') as Container;
>         service.connections.allowFromAnyIpv4(
>           ec2.Port.tcp(container.trafficPort),
>           'Accept inbound traffic on traffic port from anywhere',
>         );
>       }
>     }
>   ```
>   
>   
>   In keeping with the spirit of default config that works out of the box, I think the extension should open the SG itself if you have specified a DNS zone

Sounds good to me. I'll make this change and push it up when it's ready.

> * If you change the DNS zone property after deploying the service (as in create the service with record in one DNS zone, and then change the dns zone it seems to get confused (to be expected though, tbh). I don't think this is worth investing a ton of effort into as its an extreme edge case, but we should perhaps add a comment in the doc string for the extension indicating that you will want to remove the extension completely and then readd it if you want to change the DNS zone.

Good catch! I can see why this happens too. But now I'm thinking that if the record name changes, we'll have the same problem - that might be a little less extreme an edge case. I have a few ideas on how to address this. I'll experiment and get back to you on this.

> * This is the most important issue. If the hosted zone has multiple records it sometimes overwrites the wrong one. See in this example the Lambda was configured to write to `name.nathanpeck.gg`
>   ![Screen Shot 2020-10-09 at 3 04 55 PM](https://user-images.githubusercontent.com/3495674/95622156-e91b5780-0a40-11eb-9086-c7b06fe95915.png)
>   However when it actually executed it selected another prexisting record in the same hosted zone: `stats.nathanpeck.gg` and overwrite it instead of writing to `name.nathanpeck.gg`
>   ![Screen Shot 2020-10-09 at 3 04 42 PM](https://user-images.githubusercontent.com/3495674/95622203-00f2db80-0a41-11eb-80a1-833a6471373a.png)
>   I believe that these lines right here need to be modified to select the correct matching record, rather than just the first record: https://github.com/aws/aws-cdk/pull/10646/files#diff-597cb6e28116f127342830001126eb7dR46-R48

Hmm. Interesting and good catch! I was expecting `ListResourceRecordSets` to return only the record in question given the parameters. I think I see the mistake. :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 72387145c33879d7936387be99509195bfb88cbc
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bc68e83b-3202-4467-b534-aad5e934ab13%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cd8589795a96444dcf40589813dbfd5e9bffff16
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f46437e1-4a70-4199-8f0a-89f36a95e2ff%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@nathanpeck I believe I've more or less addressed your observations in the latest commits.

The one thing nagging at me is the fact that the system becomes confused when we change dns info. I was thinking that we could tighten this part up if we change the partition key of the table to be a composite of the cluster arn and service name rather than zone id and record name. Then, at least the lambda wouldn't ever be confused about the IPs belonging to the service.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8d7ceb9f938684a445d3d1f77bcba29c6aaa995a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fcaa7d9c-b458-47fe-9c38-6f138096f0c5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f08515231214496feadea40058bb56acdf704dc8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=368fd176-cec4-4307-a2f8-a9ee242028f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nathanpeck commented:
Looks good to me! I'll track down a member with approval access to get this merged!

misterjoshua commented:
> Looks good to me! I'll track down a member with approval access to get this merged!

Sounds good. I'll keep an eye out for the feedback. :)

misterjoshua commented:
> Okay looks like we are merging in now, and this should go out in the next CDK build! Thanks a lot for the excellent contribution @misterjoshua

Great news!  It was my pleasure.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ca718482be9e54d22091cdaecea663ef6de7b049
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f892f6a0-0fe9-496b-8a3c-265c096dec03%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

deleugpn commented:
The lengths that we go just because ECS Service Discovery only works with private IPs is astonishing.
